README.md
# dwd12
DWD12 - Password generator based in Diceware

Diceware é um método de criação de senhas elaborado por Arnold Reinhold e publicado em 1995. Nesse método, você sorteia algumas palavras (geralmente entre 4 e 8 palavras) a partir de uma lista enorme (uma lista de 7.776 palavras para ser exato) utilizando dados comuns, de seis faces, e esta é sua senha. Mesmo a lista total sendo pública, a chance de ser quebrada é pequena, pois é de 1 em x ^ y, sendo x o total de palavras da lista e y o número de sorteios. A chance de adivinhar uma senha de 5 palavras em uma tentativa, por exemplo, é de 1 em 7.776 ^ 5, o que dá 1 em 28.430.288.029.929.701.376! Ou aproximadamente 1 em 2,8 * 10 ^ 19! Cada palavra é sorteada a partir de 5 rolagens de dados. Você pode entender melhor a proposta lendo o artigo de Micah Lee para The Intercept, entitulado Senhas fáceis para você memorizar e que nem a NSA conseguirá desvendar.

A proposta Diceware D12 (ou DW12) é de uma variante do Diceware tradicional, com algumas diferenças marcantes:

* Uso de dados de 12 faces. Dados de 12 faces oferecem individualmente um número bom de possibilidades. Podem ser substituídos por 2 rolagens de um dado de 6 faces ou mesmo por um baralho após uma adaptação rápida. Embora incomuns, estão presentes em qualquer “conjunto de dados para RPG”.
* Ao invés de uma lista única de palavras, o DW12 utiliza o conceito de tomos. Antes de sortear, você escolhe tomos de onde será feito o sorteio. Cada tomo contém 1.728 palavras, de modo que o uso de 5 tomos já supera o Diceware tradicional em entropia (totaliza 8.640 palavras). Sem contar que para tentar quebrar sua senha, mesmo que o atacante conheça o método DW12, ele não saberá que subconjunto de tomos você utilizou. E a senha se torna ainda mais difícil se utilizamos um “tomo secreto”.

Este projeto fornecerá maneira automatizada de gerar senhas e criar/gerenciar tomos.


CHANGELOG.txt
0.1    2020-01-02: Created git project
0.2    2020-01-02: Added vols [inicial]: estados brasileiros, machado de assis, pimentel, queiros and quental
0.3    2020-01-02: Added secret/special volume: asteorids.txt
0.4    2020-01-02: Inicial dwd12.sh file. Working without any customization
0.5    2020-01-02: dwd12 with -s option, to set DWD12 set.
0.6    2020-01-02: dwd12 using Volume Set paths
0.7    2020-01-02: dwd12 with -w option, to set passphrase size (in number of words)
0.8    2020-01-02: dwd12 -i shows information about the volumes set.
0.9    2020-01-02: dwd12 -v to verbose
0.10   2020-01-02: using the number of volumes in set.
0.11   2020-01-02: dwd12 -l lists all available sets.
0.12   2020-01-03: dwd12 -X lists all secret volumes available.
0.13   2020-01-03: dwd12 -x _sec set the secret file to use (not using yet)
0.14   2020-01-03: dwd12 working with secret volume.
0.15   2020-01-03: dwd12 -z _sec set the secret file and force 1 use.
0.16   2020-01-03: dwd12 fixed bug - error when duplicate volumes available.
0.17   2020-01-03: dwd12 -i shows the list of volumes in choosen set.
0.18   2020-01-09: dwd12: fixed bug in -x option.
0.19   2020-01-09: dwd12: using configuration file (not working yet)
0.20   2020-01-10: dwd12: fixed config-file reading
0.21   2020-01-10: added volume set "fast" (7 vols)
A0.1   2020-01-21: dwd12admin: concept
A0.2   2020-02-21: dwd12admin: -d to set destination (global, local or user)
A0.3   2020-02-22: dwd12admin: -n to set destination name
A0.4   2020-02-22: dwd12admin: -c to copy a set to another name/local
A0.5   2020-02-22: dwd12admin: -x to copy a common volume to be a secret volume
A0.6   2020-02-27: dwd12admin: -i to install a set in current directory
A0.7   2022-11-05: dwd12: documentation for shdoc
A0.8   2022-11-05: dwd12 translated to pt_BR (using gettext)
A0.9   2022-11-05: Makefile fixed to install pt_BR translation
A0.10  2022-11-05: 3 secret volumes added: multiverso[123]
A0.11  2022-11-09: Makefile with user-sets command
A0.12  2022-12-09: dwd12admin documented with shdoc
A0.13  2022-12-16: dwd12admin -t _file to test/check a volume
A0.14  2022-12-16: dwd12admin -T _dir, and -t _file -T _dir
A0.15  2022-12-16: fixed volumes in inicial set
A0.16  2022-12-16: fixed volumes in fast set
A0.17  2022-12-16: dwd12admin -w to extract words (helping to create new volumes)
A0.18  2022-12-18: fixed sume bugs in dwd12.sh
A0.19  2023-07-23: Created hotfile to build a "codeberg-page" with 2hot


dwd12admin.sh
#!/usr/bin/env bash
# @file dwd12admin.sh
# @brief DiceWare D12 admin tools
# @description
# Documentation for shdoc - https://github.com/reconquest/shdoc

DWD12VERSION='A0.17'

GLOBALSETS=/usr/lib/dwd12/sets
LOCALSETS=/usr/local/lib/dwd12/sets
USERSETS=$HOME/.dwd12/sets

GLOBALSECS=/usr/lib/dwd12/secrets
LOCALSECS=/usr/local/lib/dwd12/secrets
USERSECS=$HOME/.dwd12/secrets

DWD12LOCALE=''
for AUX in $HOME/.dwd12 /usr/local/lib/dwd12 /usr/lib/dwd12 .
do
  if [ -d "$AUX/sets" ]
  then
    DWD12VOLS="$AUX/sets $DWD12VOLS"
  fi
  if [ -d "$AUX/locale" ]
  then
    DWD12LOCALE="$AUX/locale"
  fi
done

destination='user'
name='undefined'

_verbose=0 #Disable
_vfile=$(mktemp)

# Gettext configure
source gettext.sh

# @description Prints message in user language, using Gettext
# @arg $1 string Message in English to translate
_1text() {
	TEXTDOMAINDIR="$DWD12LOCALE" gettext 'DWD12' "$*"
}

# @description Print only if mode verbose is active
# @arg $1 string String to Print
_vprint() {
  echo "${FUNCNAME[1]} $1" >> $_vfile
}

# @description Show global variables, if verbose
# @arg $1 string Text to show that defines the moment of the code
# @stdout Introductory text and variable values
_showvars() {
  if [ $_verbose -eq 2 ]
  then
    echo $1
    printf "%s:\n" "$(_1text "Variables")"
    printf "  %s: %s\n" "$(_1text "Destination")" "$destination"
    printf "  %s: %s\n" "$(_1text "Name")" "$name"
    printf "  %s: %s\n" "$(_1text "Verbose")" "$_verbose"
  fi
}

# @description Set verbose mode on/off
# @arg $1 string Text "true" or "false"
_setverbose() {
  _showvars "$(_1text "Before set verbose")"
  case $1 in
    false )
      _verbose=0
      _vprint "$(_1text "Mode verbose turned off")"
      ;;
    true )
      _verbose=1
      _vprint "$(_1text "Mode verbose turned on")"
      ;;
  esac
  _showvars "$(_1text "After set verbose")"
}

# @description Copy a file to global, local or user directory
# @arg $1 string Origin name
# @arg $2 string Destiny name
generic_copy() {
  local FULLOR ORNAME
  FULLOR="$1"
  ORNAME="$2"
  case "$destination" in
    global )
      FULLDE="$GLOBALSETS/$name"
      mkdir -p "$GLOBALSETS"
      ;;
    local )
      FULLDE="$LOCALSETS/$name"
      mkdir -p "$LOCALSETS"
      ;;
    user )
      FULLDE="$USERSETS/$name"
      mkdir -p "$USERSETS"
      ;;
    * )
      exit;
      ;;
  esac
  if [ -d "$FULLDE" ]
  then
    printf "$(_1text "Set %s yet created in %s!")\n" $name $FULLDE
  else
    _vprint "cp -R $FULLOR $FULLDE"
    cp -R "$FULLOR" "$FULLDE"
  fi
}

# @description Copy full set
# @arg $1 string Set origin
dcopy() {
  local ORIGIN FULLOR
  ORIGIN="$1"
  if [ "$name" = "undefined" ]
  then
    _1text "You need to use -n name to set the name of the copy."
    echo
    exit
  fi
  FULLOR="$USERSETS/$ORIGIN"
  if [ ! -d "$FULLOR" ]
  then
    FULLOR="$LOCALSETS/$ORIGIN"
    if [ ! -d "$FULLOR" ]
    then
      FULLOR="$GLOBALSETS/$ORIGIN"
      if [ ! -d "$FULLOR" ]
      then
        printf "$(_1text "Set %s not found.")\n" $ORIGIN
        exit
      fi
    fi
  fi
  generic_copy "$FULLOR" "$name"
}

# @description Copy a volume from set to a new secret
# @arg $1 string Secret volume origin in form: Set/volume
scopy() {
  local SETNAM VOLNAM FULLOR FULLDE
  SETNAM=$(echo $1 | cut -d\/ -f 1)
  VOLNAM=$(echo $1 | cut -d\/ -f 2)
  if [ "$name" = "undefined" ]
  then
    name="$VOLNAM"
    exit
  fi
  case "$destination" in
    global )
      FULLDE="$GLOBALSECS/$name.txt"
      mkdir -p "$GLOBALSECS"
      ;;
    local )
      FULLDE="$LOCALSECS/$name.txt"
      mkdir -p "$LOCALSECS"
      ;;
    user )
      FULLDE="$USERSECS/$name.txt"
      mkdir -p "$USERSECS"
      ;;
    * )
      exit;
      ;;
  esac
  FULLOR="$USERSETS/$SETNAM"
  if [ ! -d "$FULLOR" ]
  then
    FULLOR="$LOCALSETS/$SETNAM"
    if [ ! -d "$FULLOR" ]
    then
      FULLOR="$GLOBALSETS/$SETNAM"
      if [ ! -d "$FULLOR" ]
      then
        printf "$(_1text "Set %s not found.")\n" $SETNAM
        exit
      fi
    fi
  fi
  FULLOR="$FULLOR/$VOLNAM.txt"
  if [ ! -f "$FULLOR" ]
  then
    printf "$(_1text "Origin volume %s not found in set %s (%s).")\n" "$VOLNAM" "$SETNAM" "$FULLOR"
    exit
  fi
  if [ -f "$FULLDE" ]
  then
    printf "$(_1text "Secret volume named %s exists in %s!")\n" $name $FULLDE
  else
    _vprint "cp $FULLOR $FULLDE"
    cp "$FULLOR" "$FULLDE"
  fi
}

# @description Install a set from current directory
# @arg $1 string Origin set
setinstall() {
  local ORIGIN
  ORIG="$1"
  if [ "$name" = "undefined" ]
  then
    name=$(basename "$ORIG")
  fi
  if [ -d "$ORIG" ]
  then
    _vprint "Installing $ORIG to $destination."
    generic_copy "$ORIG"
  else
    printf "$(_1text "Set %s not found in current directory (%s).")\n" "$ORIG" "$(pwd)"
  fi
}

# @description Check a volume or set
# @arg $1 string Filename or dirname
dwcheck() {
  if [ -f "$1" ]
  then
    dwcheck.volume "$1"
  elif [ -d "$1" ]
  then
    dwcheck.set "$1"
  else
    printf "$(_1text "No volume or set found for %s.")\n" "$1"
  fi
}

# @description Check a volume and prints some stats about it
# @arg $1 string Filename
dwcheck.volume() {
  local _file _chars _words _lines _dupls _maxim _aux _avera _cents
  local _wchk _lchk
  _file="$1"  
  _lines=$(cat "$_file" | wc -l)
  if [ $_lines -eq 1728 ]
  then
    _lchk="$(_1text "ok")"
  else
    _lchk="$(_1text "wrong")"
  fi
  _words=$(cat "$_file" | wc -w)
  if [ $_words -eq 1728 ]
  then
    _wchk="$(_1text "ok")"
  else
    _wchk="$(_1text "wrong")"
  fi
  _chars=$(cat "$_file" | wc -c)
  _maxim=$(cat "$_file" | wc -L)
  _aux=$(( (_chars - _lines)*10/_lines))
  _avera=$((_aux/10))
  _cents=$((_aux%10))
  _dupls=$(cat "$_file" | sort | uniq -d | wc -w)
  echo
  printf "$(_1text "File checked: %s")\n" "$_file"
  if [ $_dupls -eq 0 ]
  then
    printf "$(_1text "Duplicates: 0 [%s]")\n" "$(_1text "ok")"
  else
    printf "$(_1text "Duplicates: %i [%s]")\n" \
      $_dupls "$(cat "$_file" | sort | uniq -d | xargs)"
  fi
  printf "$(_1text "Words: %i [%s]")\n" $_words $_wchk
  printf "$(_1text "Lines: %i [%s]")\n" $_lines $_lchk
  printf "$(_1text "Chars per word: %i.%i")\n" $_avera $_cents
  printf "$(_1text "Greater word: %i chars")\n" $_maxim
  if [ $_words -eq 1728 -a $_lines -eq 1728 -a $_dupls -eq 0 ]
  then
    _1text "This volume is fine."
    echo
    return 0
  else
    _1text "You need to fix some things."
    echo
    return 1
  fi
}

# @description Test a volume set
# @arg $1 string Set dir name
dwcheck.set() {
  local _file _errors _vols
  pushd $1 > /dev/null
  _errors=0
  for _file in $(ls *.txt)
  do
    dwcheck.volume "$_file"
    if [ $? -gt 0 ]
    then
      _errors=$((_errors+1))
    fi
  done
  _vols=$(ls *.txt | wc -l)
  popd > /dev/null
  echo
  printf "$(_1text "Volumes: %i")\n" $_vols
  printf "$(_1text "Volumes to fix: %i")\n" $_errors
}

# @description Test a volume comparing with all tomes of some set
# @arg $1 string File names
dwcheck.full() {
  local _dupls
  echo
  printf "$(_1text "BIG DUPLICATE TEST [%s]")\n" "$*"
  _dupls=$(cat $* | sort | uniq -d | wc -w)
  if [ $_dupls -eq 0 ]
  then
    _1text "Full test: passed"
    echo
  else
    _1text "Duplicated words:"
    echo
    cat $* | sort | uniq -d
  fi
}

# @description Check if word is in a set or volume
# @arg $1 string Word to check
# @arg $2 string Set or volume
dwextract.check() {
  local _i
  if [ -f "$2" ]
  then
    grep -q "^$1\$" "$2"
    return $?
  else
    for _i in $(ls $2/*.txt)
    do
      grep "^$1\$" "$_i"
      if [ $? -eq 0 ]
      then
        return 1
      fi
    done
    return 0
  fi
}

# @description Extract words from a text file
# @arg $1 string File from where to extract
# @arg $2 string File or dir to compare (optional)
dwextract() {
  local wrd wok wno
  local tmp=$(mktemp)
  wok=0
  wno=0
  cat "$1" | xargs | tr ' ' '\n' > $tmp
  case $# in
    1)
      cat $tmp
      ;;
    2)
      for wrd in $(cat $tmp)
      do
        dwextract.check "$wrd" "$2"
        if [ $? -eq 0 ]
        then
          echo $wrd
          wok=$((wok+1))
        else
          wno=$((wno+1))
        fi
      done
      printf "$(_1text "New words: %i")\n" $wok
      printf "$(_1text "Repeated words: %i")\n" $wno
      ;;
  esac
  rm $tmp
}

# @description Print help menu
# @stdout Usage instructions
showhelp() {
  echo "DWD12 Admin $DWD12VERSION"
  echo "      $(_1text "Manage your DWD12 volumes!")"
  echo
  echo "$ dwd12admin [$(_1text "options")]"
  echo
  echo "  -d _dest  $(_1text "Destination (global, local or user)")"
  echo "  -n _new   $(_1text "Destination name")"
  echo "  -c _set   $(_1text "Copy this set")"
  echo "  -x _s/_v  $(_1text "Copy this volume to a secret (setname/volname)")"
  echo "  -i _set   $(_1text "Install this set (directory)")"
  echo "  -I _svol  $(_1text "Install this secret volume [TO DO]")"
  echo "  -u _set   $(_1text "Uninstall this set [TO DO]")"
  echo "  -U _svol  $(_1text "Uninstall this secret volume [TO DO]")"
  echo "  -V _set   $(_1text "List all volumes in a set. [TO DO]")"
  echo "  -t _file  $(_1text "Test a local volume.")"
  echo "  -T _dir   $(_1text "Test a local volume set.")"
  echo "  -w _file  $(_1text "Extract words from a file.")"
  echo "  -s        $(_1text "List all volume sets. [TO DO]")"
  echo "  -S        $(_1text "List all secret volumes [TO DO]")"
  echo "  -v        $(_1text "Enable verbose mode.")"
  echo "  -h        $(_1text "Show this help menu")"
  echo
}

tdir="" # test dir
tfil="" # test file
xfil="" # extract words
while getopts "d:n:c:i:x:t:T:w:vh" option
do
  case ${option} in
    d  )
      case "$OPTARG" in
        global|local|user)
          destination="$OPTARG"
          _vprint "$(printf "$(_1text "Setting destination to %s")" "$destination")"
          ;;
        *)
          echo "$(_1text "Unknown option! Set destination to global, local or user.")"
          ;;
      esac
      ;;
    n )
      name="$OPTARG"
      _vprint "$(printf "$(_1text "Setting destination name to %s")" "$name")"
      ;;
    c )
      dcopy "$OPTARG"
      ;;
    i )
      setinstall "$OPTARG"
      ;;
    x )
      scopy "$OPTARG"
      ;;
    t )
      tfil="$OPTARG"
      ;;
    T )
      tdir="$OPTARG"
      ;;
    w )
      wfil="$OPTARG"
      ;;
    v ) #Set mode verbose
      _setverbose "true"
      ;;
    h | \? )
      showhelp
      rm $_vfile
      exit
      ;;
  esac
done

if [ "$tfil" != "" ] && [ "$tdir" != "" ]
then
  dwcheck "$tdir"
  dwcheck "$tfil"
  dwcheck.full $tfil $tdir/*.txt
elif [ "$wfil" != "" ] && [ "$tfil" != "" ]
then
  dwextract "$wfil" "$tfil"
elif [ "$tfil" != "" ]
then
  dwcheck "$tfil"
elif [ "$tdir" != "" ]
then
  dwcheck "$tdir"
elif [ "$wfil" != "" ]
then
  dwextract "$wfil"
fi

if [ $_verbose -gt 0 ]
then
  cat $_vfile
fi
rm $_vfile


dwd12.sh
#!/bin/bash
# @file dwd12.sh
# @brief DiceWare D12 password generator
# @description
# Documentation for shdoc - https://github.com/reconquest/shdoc

DWD12VERSION='A0.18'

# Gettext configure
source gettext.sh

# @description Prints message in user language, using Gettext
# @arg $1 string Message in English to translate
_1text() {
	TEXTDOMAINDIR="$DWD12LOCALE" gettext 'DWD12' "$*"
}

DWD12VOLS=''
DWD12LOCALE=''
for AUX in $HOME/.dwd12 /usr/local/lib/dwd12 /usr/lib/dwd12 .
do
  if [ -d "$AUX/sets" ]
  then
    DWD12VOLS="$AUX/sets $DWD12VOLS"
  fi
  if [ -d "$AUX/locale" ]
  then
    DWD12LOCALE="$AUX/locale"
  fi
done
if [ "$DWD12VOLS" == "" ]
then
  _1text "Error! No DWD12 volumes path created."
  echo
  exit
fi
if [ "$DWD12LOCALE" == "" ]
then
  DWD12LOCALE="locale/"
fi

DWD12SECRETS=''
for AUX in $HOME/.dwd12/secrets /usr/local/lib/dwd12/secrets /usr/lib/dwd12/secrets ./secrets
do
  if [ -d "$AUX" ]
  then
    DWD12SECRETS="$AUX $DWD12SECRETS"
  fi
done
if [ "$DWD12SECRETS" == "" ]
then
  _1text "Error! No DWD12 secrets path created."
  echo
  exit
fi
DWD12SIZE=4
_mode="normal"
_verbose=0 #Disable
_vfile=$(mktemp)

# @description Print only if mode verbose is active
# @arg $1 string Message to print
_vprint() {
  echo "${FUNCNAME[1]} $*" >> $_vfile
}

# @description Run x d12
# @arg $1 int Number of dices. Default: 1
_rund12() {
  local DICES
	if [ $# -eq 0 ]
	then
		DICES=1
	else
		DICES="$1"
	fi
  _vprint "$(printf "$(_1text "Roll %sd12")" "$DICES")"
	tr -dc '123456789XJQ' < /dev/urandom | head -c "$DICES" | sed 's/\(.\)/\1\  /g' | sed 's/X/10/g' | sed 's/J/11/g' | sed 's/Q/12/g'; echo
}

# @description Sort a word
# @arg $1 int Total of volumes
_sortaword() {
  local VOLS MV PLUS V2 WRD VOL MSG
	VOLS=$_dvls
	if [ "$#" -ge 1 ]
	then
		if [ "$1" -gt 0 ]
		then
			if [ "$1" -lt 20 ]
			then
				VOLS="$1"
			fi
		fi
	fi
  _vprint "$(printf "$(_1text "Will work with %i volumes.")" "$VOLS")"
	MV=""
	PLUS=""
	if [ "$VOLS" -ge 10 ]
	then
		(( "V2 = $VOLS - 10" ))
		if [ "$V2" -ge 10 ]
		then
			_1text "It don't works for more then 19 volumes."
      echo
			return 1
		fi
		PLUS="A"
		if [ "$V2" -ge 1 ]; then PLUS="$PLUS""B"; fi
		if [ "$V2" -ge 2 ]; then PLUS="$PLUS""C"; fi
		if [ "$V2" -ge 3 ]; then PLUS="$PLUS""D"; fi
		if [ "$V2" -ge 4 ]; then PLUS="$PLUS""E"; fi
		if [ "$V2" -ge 5 ]; then PLUS="$PLUS""F"; fi
		if [ "$V2" -ge 6 ]; then PLUS="$PLUS""G"; fi
		if [ "$V2" -ge 7 ]; then PLUS="$PLUS""H"; fi
		if [ "$V2" -ge 8 ]; then PLUS="$PLUS""I"; fi
		if [ "$V2" -ge 9 ]; then PLUS="$PLUS""J"; fi
	fi
	if [ "$VOLS" -ge 1 ]; then MV="1"; fi
	if [ "$VOLS" -ge 2 ]; then MV="$MV""2"; fi
	if [ "$VOLS" -ge 3 ]; then MV="$MV""3"; fi
	if [ "$VOLS" -ge 4 ]; then MV="$MV""4"; fi
	if [ "$VOLS" -ge 5 ]; then MV="$MV""5"; fi
	if [ "$VOLS" -ge 6 ]; then MV="$MV""6"; fi
	if [ "$VOLS" -ge 7 ]; then MV="$MV""7"; fi
	if [ "$VOLS" -ge 8 ]; then MV="$MV""8"; fi
	if [ "$VOLS" -ge 9 ]; then MV="$MV""9"; fi
	MV="$MV$PLUS"
	WRD=$(_rund12 3)
  
	VOL=$(tr -dc "$MV" < /dev/urandom | head -c 1 | sed 's/^//g'; echo)
	if [ "$VOL" = 'A' ]; then VOL=10; fi
	if [ "$VOL" = 'B' ]; then VOL=11; fi
	if [ "$VOL" = 'C' ]; then VOL=12; fi
	if [ "$VOL" = 'D' ]; then VOL=13; fi
	if [ "$VOL" = 'E' ]; then VOL=14; fi
	if [ "$VOL" = 'F' ]; then VOL=15; fi
	if [ "$VOL" = 'G' ]; then VOL=16; fi
	if [ "$VOL" = 'H' ]; then VOL=17; fi
	if [ "$VOL" = 'I' ]; then VOL=18; fi
	if [ "$VOL" = 'J' ]; then VOL=19; fi
  MSG="$(printf "$(_1text "Volume %i, %s")" $VOL "$WRD")"
  _vprint "$MSG"
  echo "$MSG"
}

# @description Sort the passphrase, just positions
# @arg $1 int Number of volumes (default: 4)
# @arg $2 int Number of words (default: 4)
_predwd12() {
  local VOLS MSG

	VOLS=$_dvls
	if [ "$#" -ge 1 ]
	then
		if [ "$1" -gt 0 ]
		then
			if [ "$1" -lt 20 ]
			then
				VOLS="$1"
			fi
		fi
	fi
	for i in $(seq 1 "$DWD12SIZE")
	do
		MSG="$(printf "$(_1text "WORD %i: %s")" $i "$(_sortaword "$VOLS")")"
    _vprint "$MSG"
    echo "$MSG"
	done
}

# @description Password generator
rundwd12() {
  local ACTPOS VOLS SECPOS j IJ
  ACTPOS=0
  _vprint "$(printf "$(_1text "Running with %s set. Generating passphrase with %i words.")" "$DWD12SET" $DWD12SIZE)"
  
  if [ "$_mode" == "secure" ]
  then
    VOLS=$_dvls
    SECPOS=$(shuf -i 1-$DWD12SIZE -n 1)
    _vprint "$(printf "$(_1text "The position %i will turn into a secret-vol word.")" "$SECPOS")"
  else
    VOLS=$((_dvls + _dsec))
    SECPOS=-1
    if [ $_dsec -eq 1 ]
    then
      _vprint "$(printf "$(_1text "There are %i volumes. Number %i is the secret (%s)!")" $_dvls $VOLS "$DWD12SEC")"
    fi
  fi

  if [ $_verbose -eq 1 ]
  then
    _1text "Passphrase: "
  fi
  
  for j in $(_predwd12 "$VOLS" | \
      sed 's/[[:alpha:]]//g' | \
      sed 's/^ //' | \
      sed 's/://' | \
      sed 's/,//' | \
      sed 's/[ ]\+/-/g')
	do

		#IJ is an array from word shuffle result from dwd12
		# 0: word counter; 1: vol; 2, 3, 4: words
		IJ=(${j//-/ })
    ACTPOS=$((ACTPOS + 1))
    if [ $_dsec -eq 1 ]
    then
      # If mode is normal, x==max+1 => special
      # If mode is secure, secpos is a random number in (1..max)
      if [ ${IJ[1]} -eq $((_dvls + 1)) -o $ACTPOS -eq $SECPOS ]
      then
        JFILE=$(find $DWD12SECRETS -maxdepth 1 -name "$DWD12SEC.txt" | head -1)
      else
		    JFILE=$(find "$_dset" -maxdepth 1 -type f | sed -n "${IJ[1]}p")
      fi
    else
      JFILE=$(find "$_dset" -maxdepth 1 -type f | sed -n "${IJ[1]}p")
    fi
    
    _vprint "$(printf "$(_1text "Using a word from volume %s.")" "$JFILE")"
		JWORD=$(( (IJ[2]-1)*144+(IJ[3]-1)*12+IJ[4] ))
		JDW=$(sed -n "$JWORD"p < "$JFILE")
		echo -n "$JDW "
	done
	echo
}

# @description Choose a DWD12 set
# @arg $1 string Set ID
_dwd12set() {
  DWD12SET="$1"
  _vprint "$(printf "$(_1text "Selected set %s")" "$DWD12SET")"
}

# @description Choose the secret DWD12 volume
# @arg $1 string Volume name
_dwd12secret() {
  if [ $(find $DWD12SECRETS -mindepth 1 -maxdepth 1 -name $1.txt | head -1) != "" ]
  then
    _dsec=1
    DWD12SEC="$1"
    _vprint "$(printf "$(_1text "Selected secret volume %s")" "$DWD12SEC")"
  else
    printf "$(_1text "Error! Secret volume %s not found.")\n" "$1"
    rm $_vfile
    exit
  fi
}

# @description Set passphrasesize
# @arg $1 int Passwphrase size
_dwd12words() {
  if [ "$1" -gt 0 ]
  then
    if [ "$1" -lt 50 ]
    then
      DWD12SIZE="$1"
      _vprint "$(printf "$(_1text "Passphrase size set to %i")" $DWD12SIZE)"
    fi
  fi
}

# @description Set verbose mode
# @arg $1 string false or true
_dwd12verbose() {
  _showvars "Before set verbose"
  case $1 in
    false )
      _verbose=0
      _vprint "$(_1text "Mode verbose turned off")"
      ;;
    true )
      _verbose=1
      _vprint "$(_1text "Mode verbose turned on")"
      ;;
  esac
  _showvars "After set verbose"
}

# @description Prints all variable and values
_showvars() {
  if [ $_verbose -eq 2 ]
  then
    echo $1
    printf "$(_1text "Variables:")\n"
    printf "  $(_1text "Set: %s")\n" "$DWD12SET"
    printf "  $(_1text "Secret: %s")\n" "$DWD12SEC"
    printf "  $(_1text "Mode: %s")\n" "$_mode"
    printf "  $(_1text "Size: %i words")\n" $DWD12SIZE
    printf "  $(_1text "Verbose: %s")\n" "$_verbose"
  fi
}

# @description List all available volumes sets
showsets() {
  local s
  for s in $(find $DWD12VOLS -mindepth 1 -maxdepth 1 -type d)
  do
    echo $(basename $s) "  $s"
  done
}

# @description List all available secret volumes
showsecrets() {
  local s
  for s in $(find $DWD12SECRETS -maxdepth 1 -name '*.txt')
  do
    echo $(basename $s .txt) "  $s"
  done
}

# @description Prints information about actual vomules set
showinfo() {
  local i
  echo "DWD12 $DWD12VERSION"
  echo $0
  echo
  printf "$(_1text "Set: %s")\n" "$DWD12SET"
  printf "$(_1text "Path: %s")\n" "$_dset"
  printf "$(_1text "Volumes: %s")\n" "$_dvls"
  for i in $(find "$_dset" -maxdepth 1 -type f)
  do
    echo "  -" $(basename $i)
  done
  echo

  if [ $_dsec == 0 ]
  then
    printf "$(_1text "Not using secret volume.")\n"
  else
    printf "$(_1text "Secret volume: %s")\n" "$DWD12SEC"
  fi

  for i in $DWD12VOLS
  do
    printf "$(_1text "Sets in %i: %s.")\n" $i "$(ls $i | wc -l)"
  done

}

# @description Print help menu
showhelp() {
  echo "DWD12 $DWD12VERSION"
  echo "      $(_1text "Passphrase generator!")"
  echo
  echo "$ dwd12 [$(_1text "options")]"
  echo
  echo "  -s _set   $(_1text "Generate passphrase using the set _set.")"
  echo "  -x _svol  $(_1text "Use given secret volume.")"
  echo "  -z _svol  $(_1text "Use given secret volume (exactly 1 of the words in passphrase).")"
  echo "  -w _size  $(_1text "Gerenate passphrase with _size words.")"
  echo "  -i        $(_1text "Just show information about selected set.")"
  echo "  -l        $(_1text "List all volumes sets available to use.")"
  echo "  -X        $(_1text "List all secret volumes available.")"
  echo "  -v        $(_1text "Enable verbose mode.")"
  echo "  -h        $(_1text "Show this help menu")"
  echo
}

# @description Read the configuration file. PENDING
_readconf() {
  local FILE RCPARAM RCVALUE
  FILE="$1"
  if [ -f "$FILE" ]
  then
    for line in $(egrep -v '^#' "$FILE" | egrep -v '^$')
    do
      RCPARAM=$(echo "$line" | cut -d '=' -f 1)
      RCVALUE=$(echo "$line" | cut -d '=' -f 2-)
      _vprint "From $FILE, $RCPARAM set to $RCVALUE"
      case $RCPARAM in
        SET ) _dwd12set "$RCVALUE" ;;
        SECRET ) _dwd12secret "$RCVALUE" ;;
        MODE )
          case "$RCVALUE" in
            normal | secure )
              _mode=$RCVALUE
              _vprint "$(printf "$(_1text "Mode set to %s")" "$RCVALUE")"
              ;;
            secret ) _mode=normal ;;
            * ) _vprint "$(printf "$(_1text "Invalid option (%s) to %s.")" "$RCVALUE" "$RCPARAM")"
          esac
          ;;
        WORDS ) _dwd12words "$RCVALUE" ;;
        VERBOSE )
          _dwd12verbose "$RCVALUE"
          ;;
      esac
    done
    _showvars "$(_1text "Inside readconf, middle.")"
  else
    _vprint "$(printf "$(_1text "Config file %s not found.")" "$FILE")"
  fi
  _showvars "$(_1text "Inside readconf, before go out.")"
}


# MAIN CODE

DWD12SET=$(showsets | head -1 | cut -d\  -f 1)
DWD12SEC=$(basename $(find secrets -name '*.txt' | head -1) .txt)
_dsec=0

_showvars "$(_1text "Before configuration file")"

_readconf "/etc/dwd12/dwd12.conf"
_readconf "$HOME/.dwd12/dwd12.conf"

_showvars "$(_1text "After configuration file")"

while getopts "s:x:z:w:lXivh" option
do
  case ${option} in
    i ) #Information about the set of volumes
      _mode="info"
      _vprint "$(_1text "Mode set to info")"
      ;;
    s ) #Set of volumes
      _dwd12set "$OPTARG"
      ;;
    x ) #Set the secret volume
        _dwd12secret "$OPTARG"
        ;;
    z ) #Set the secret volume
        _dwd12secret "$OPTARG"
        _mode="secure"
        ;;
    v ) #Set mode verbose
      _dwd12verbose "true"
      ;;
    w) #Size of passphrase (in Words)
      _dwd12words "$OPTARG"
      ;;
    l  ) #List all sets available
      _1text "Sets available: "
      showsets
      rm $_vfile
      exit

      ;;
    X  ) # List all secret volumes
      _1text "Secret volumes: "
      showsecrets
      rm $_vfile
      exit

      ;;
    h  )
      showhelp
      rm $_vfile
      exit
      ;;
    \? ) #For invalid option
      showhelp
      rm $_vfile
      exit
      ;;
  esac
done

_dset=$(find $DWD12VOLS -name $DWD12SET | head -1)
_dvls=$(find "$_dset" -maxdepth 1 -type f | wc -l)

if [ "$_dset" = "" ]
then
  printf "$(_1text "Error. Set %s not found.")" "$DWD12SET"
else
  case ${_mode} in
    info)
      showinfo
    ;;
    normal|secure)
      FINAL=$(rundwd12)
      if [ $_verbose -gt 0 ]
      then
        cat $_vfile
      fi
      echo $FINAL
      rm $_vfile
    ;;
  esac
fi
